test/test_common/logging.h

```diff
@@ -61,30 +61,103 @@ class LogRecordingSink : public Logger::SinkDelegate {
   std::vector<std::string> messages_;
 };
 
-// Validates that when stmt is executed, exactly one log message containing substr will be emitted.
-#define EXPECT_LOG_CONTAINS(loglevel, substr, stmt)                                                \
+typedef std::vector<std::pair<std::string, std::string>> ExpectedLogMessages;
+
+// Validates that when stmt is executed, log messages containing substr and loglevel will be
+// emitted. Failure message e.g.,
+//
+// Logs:
+// [2018-04-12 05:51:00.245][7290192][debug][upstream] grpc_mux_impl.cc:160] Received gRPC
+// [2018-04-12 05:51:00.246][7290192][warning][upstream] grpc_mux_impl.cc:63] Called bar
+// [2018-04-12 05:51:00.246][7290192][trace][upstream] grpc_mux_impl.cc:80] Sending foo
+//  Does NOT contain:
+//    'warning', 'Too many sendDiscoveryRequest calls for baz’
+//    'warning', 'Too man sendDiscoveryRequest calls for foo'
+#define EXPECT_LOG_CONTAINS_MANY(expected_messages, stmt)                                          \
+  do {                                                                                             \
+    ASSERT_FALSE(expected_messages.empty()) << "Expected messages cannot be empty.";               \
+    LogLevelSetter save_levels(spdlog::level::trace);                                              \
+    LogRecordingSink log_recorder(Logger::Registry::getSink());                                    \
+    stmt;                                                                                          \
+    if (log_recorder.messages().empty()) {                                                         \
+      FAIL() << "Expected message(s), but NONE was recorded.";                                     \
+    }                                                                                              \
+    ExpectedLogMessages failed_expectations;                                                       \
+    for (const auto& expected : expected_messages) {                                               \
+      const auto log_message =                                                                     \
+          std::find_if(log_recorder.messages().begin(), log_recorder.messages().end(),             \
+                       [&](const std::string& message) {                                           \
```

Comment:
Just capture '&expected', rather than capturing all the state in the enclosing function.